tsdb/index/tsi1/index.go

```diff
@@ -829,11 +819,11 @@ func (i *Index) compactionGroups(fileSet FileSet) [][]*IndexFile {
 
 		// Otherwise append to the current group.
 		group = append(group, indexFile)
-	}
 
-	// Append final group, if compactable.
-	if isCompactableGroup(group, i.CompactionFactor) {
-		groups = append(groups, group)
+		// Append group, if compactable.
+		if isCompactableGroup(group, i.CompactionFactor) {
+			group, groups = nil, append(groups, group)
+		}
```

Comment:
This section was outside the `for` loop to grab any remaining files in a compaction group.

By moving it inside the loop it is forming compaction groups as soon as a valid one exists instead of trying to find the longest contiguous group of files.